File: src/HolzShots/Input/Actions/SelectAreaCommand.vb
Imports HolzShots.ScreenshotRelated.Selection
Imports HolzShots.Composition.Command
Imports HolzShots.Threading
Imports HolzShots.Drawing
Imports HolzShots.Input.Selection

Namespace Input.Actions
    <Command("captureArea")>
    Public Class SelectAreaCommand
        Inherits CapturingCommand

        Public Overrides Async Function Invoke(parameters As IReadOnlyDictionary(Of String, String), settingsContext As HSSettings) As Task
            If parameters Is Nothing Then Throw New ArgumentNullException(NameOf(parameters))
            If settingsContext Is Nothing Then Throw New ArgumentNullException(NameOf(settingsContext))

            ' TODO: Add proper assertion
            ' Debug.Assert(ManagedSettings.EnableAreaScreenshot)
            Debug.Assert(Not AreaSelector.IsInAreaSelector)

            If Not settingsContext.EnableHotkeysDuringFullscreen AndAlso HolzShots.Windows.Forms.EnvironmentEx.IsFullscreenAppRunning() Then Return

            ' TODO: Re-add proper if condition
            'If ManagedSettings.EnableAreaScreenshot AndAlso Not AreaSelector.IsInAreaSelector Then
            If Not AreaSelector.IsInAreaSelector Then
                Dim shot As Screenshot
                Try
                    shot = Await CaptureSelection(settingsContext).ConfigureAwait(True)
                    Debug.Assert(shot IsNot Nothing)
                    If shot Is Nothing Then Throw New TaskCanceledException()
                Catch cancelled As TaskCanceledException
                    Debug.WriteLine("Area Selection cancelled")
                    Return
                End Try
                Debug.Assert(shot IsNot Nothing)
                Await ProcessCapturing(shot, settingsContext).ConfigureAwait(True)
            End If

        End Function

        Shared Async Function CaptureSelection(settingsContext As HSSettings) As Task(Of Screenshot)
            If settingsContext Is Nothing Then Throw New ArgumentNullException(NameOf(settingsContext))

            Debug.Assert(Not AreaSelector.IsInAreaSelector)

            If AreaSelector.IsInAreaSelector Then Return Nothing

            Using prio As New ProcessPriorityRequest()
                Using screen = ScreenshotCreator.CaptureScreenshot(SystemInformation.VirtualScreen)
                    Using selector As New AreaSelector2(settingsContext)
                        Dim selectedArea = Await selector.PromptSelectionAsync(screen).ConfigureAwait(True)

                        Debug.Assert(selectedArea.Width > 0)
                        Debug.Assert(selectedArea.Height > 0)

                        Dim selectedImage As New Bitmap(selectedArea.Width, selectedArea.Height)

                        Using g As Graphics = Graphics.FromImage(selectedImage)
                            g.DrawImage(screen, New Rectangle(0, 0, selectedArea.Width, selectedArea.Height), selectedArea, GraphicsUnit.Pixel)
                        End Using

                        Return Screenshot.FromImage(selectedImage, Cursor.Position, ScreenshotSource.Selected)
                    End Using
                End Using
            End Using
        End Function
    End Class
End Namespace

File: src/HolzShots/Input/Action/SelectAreaCommand.cs
using System.Diagnostics;
using System.Drawing;
using System.Windows.Forms;
using HolzShots.Composition.Command;
using HolzShots.Drawing;
using HolzShots.Input.Selection;
using HolzShots.Threading;

namespace HolzShots.Input.Actions
{
    [Command("captureArea")]
    public class SelectAreaCommand : CapturingCommand
    {
        protected override async Task InvokeInternal(IReadOnlyDictionary<string, string> parameters, HSSettings settingsContext)
        {
            if (parameters == null)
                throw new ArgumentNullException(nameof(parameters));
            if (settingsContext == null)
                throw new ArgumentNullException(nameof(settingsContext));

            // TODO: Add proper assertion
            // Debug.Assert(ManagedSettings.EnableAreaScreenshot)
            Debug.Assert(!SelectionSemaphore.IsInAreaSelection);

            if (!settingsContext.EnableHotkeysDuringFullscreen && HolzShots.Windows.Forms.EnvironmentEx.IsFullscreenAppRunning())
                return;

            // TODO: Re-add proper if condition
            // If ManagedSettings.EnableAreaScreenshot AndAlso Not SelectionSemaphore.IsInAreaSelection Then
            if (!SelectionSemaphore.IsInAreaSelection)
            {
                Screenshot? shot;
                try
                {
                    shot = await CaptureSelection(settingsContext).ConfigureAwait(true);
                    Debug.Assert(shot != null);
                    if (shot == null)
                        throw new TaskCanceledException();
                }
                catch (TaskCanceledException cancelled)
                {
                    Debug.WriteLine("Area Selection cancelled");
                    return;
                }
                Debug.Assert(shot != null);
                await ProcessCapturing(shot, settingsContext).ConfigureAwait(true);
            }
        }

        public static async Task<Screenshot?> CaptureSelection(HSSettings settingsContext)
        {
            if (settingsContext == null)
                throw new ArgumentNullException(nameof(settingsContext));

            Debug.Assert(!SelectionSemaphore.IsInAreaSelection);

            if (SelectionSemaphore.IsInAreaSelection)
                return null;

            using (ProcessPriorityRequest prio = new ProcessPriorityRequest())
            {
                using (var screen = ScreenshotCreator.CaptureScreenshot(SystemInformation.VirtualScreen))
                {
                    using (var selector = AreaSelector.Create(screen, settingsContext))
                    {
                        var selectedArea = await selector.PromptSelectionAsync().ConfigureAwait(true);

                        Debug.Assert(selectedArea.Width > 0);
                        Debug.Assert(selectedArea.Height > 0);

                        var selectedImage = new Bitmap(selectedArea.Width, selectedArea.Height);

                        using var g = Graphics.FromImage(selectedImage);

                        g.DrawImage(screen, new Rectangle(0, 0, selectedArea.Width, selectedArea.Height), selectedArea, GraphicsUnit.Pixel);

                        return Screenshot.FromImage(selectedImage, Cursor.Position, ScreenshotSource.Selected);
                    }
                }
            }
        }
    }
}
